tests/test_bounce_integral.py

```diff
@@ -0,0 +1,783 @@
+            pitch, k[2:], B.c[:, 2:], B_z_ra.c[:, 2:], check=True, plot=False
+        )
+        plot_field_line(B, pitch, bp1, bp2, start=k[2])
+        bp1, bp2 = _filter_nonzero_measure(bp1, bp2)
+        assert bp1.size and bp2.size
+        intersect = B.solve(1 / pitch, extrapolate=False)
+        np.testing.assert_allclose(bp1[0], 0.835319, rtol=1e-6)
+        intersect = intersect[intersect >= k[2]]
+        np.testing.assert_allclose(bp1, intersect[[0, 2, 4]], rtol=1e-6)
+        np.testing.assert_allclose(bp2, intersect[[0, 3, 5]], rtol=1e-6)
+
+    def test_extrema_first_and_before_bp2():
+        start = -1.2 * np.pi
+        end = -2 * start + 1
+        k = np.linspace(start, end, 7)
+        B = CubicHermiteSpline(
+            k,
+            np.cos(k) + 2 * np.sin(-2 * k) + k / 10,
+            -np.sin(k) - 4 * np.cos(-2 * k) + 1 / 10,
+        )
+        B_z_ra = B.derivative()
+        pitch = 1 / B(B_z_ra.roots(extrapolate=False))[1] + 1e-13
+        # If a regression fails this test, this note will save many hours of debugging.
+        # If the filter in place to return only the distinct roots is too coarse,
+        # in particular atol < 1e-15, then this test will error. In the resulting
+        # plot that the error will produce the red bounce point on the first hump
+        # disappears. The true sequence is green, double red, green, red, green.
+        # The first green was close to the double red and hence the first of the
+        # double red root pair was erased as it was falsely detected as a duplicate.
+        # The second of the double red root pair is correctly erased. All that is
+        # left is the green. Now the bounce_points method assumes the intermediate
+        # value theorem holds for the continuous spline, so when fed these sequence
+        # of roots, the correct action is to ignore the first green root since
+        # otherwise the interior of the bounce points would be hills and not valleys.
+        bp1, bp2 = bounce_points(pitch, k, B.c, B_z_ra.c, check=True)
+        bp1, bp2 = _filter_nonzero_measure(bp1, bp2)
+        assert bp1.size and bp2.size
+        # Our routine correctly detects intersection, while scipy, jnp.root fails.
+        intersect = B.solve(1 / pitch, extrapolate=False)
+        np.testing.assert_allclose(bp1[0], -0.671904, rtol=1e-6)
+        np.testing.assert_allclose(bp1, intersect[[0, 3, 5]], rtol=1e-5)
+        # intersect array could not resolve double root as single at index 0,1
+        np.testing.assert_allclose(intersect[0], intersect[1], rtol=1e-5)
+        np.testing.assert_allclose(bp2, intersect[[2, 4, 6]], rtol=1e-5)
+
+    test_bp1_first()
+    test_bp2_first()
+    test_bp1_before_extrema()
+    test_bp2_before_extrema()
+    # In theory, this test should only pass if distinct=True when computing the
+    # intersections in bounce points. However, we can get lucky due to floating
+    # point errors, and it may also pass when distinct=False.
```

Comment:
are we relying on floating point error for this test to pass?

Reply:
no. i'll remove this comment block since it's just noise at this point